Implement readDirEntries method

vasicvuk opened the pull request:
Implementing feature to read directories from provided directory path based on Idea here:
https://github.com/helmfile/helmfile/discussions/253

Within the README.md I have also documented the readDir function which was not documented but available for usage.

yxxhero commented:
@vasicvuk Please add unittest and e2etest for this.

yxxhero commented:
@vasicvuk Thanks for your idea. I think readDirs is ambiguous. Do you have a higher function name

yxxhero commented:
@vasicvuk e2e test in `test/e2e/template/helmfile/tmpl_test.go`

vasicvuk commented:
Hi @yxxhero,

Maybe listDir or readDirsWithinDir?

Do you have some name you would like?

yxxhero commented:
@vasicvuk  listDir will better.  Can `exec` meet your needs?

vasicvuk commented:
As I understood `exec` is plarform specific and we are using helmfile on both Windows and Linux so it would be hard do handle both cases

yxxhero commented:
@mumoshu WDYT?

vasicvuk commented:
@yxxhero Maybe listDir can list both files and folders instead just folders, i think that would be better

yxxhero commented:
Can you describe your application scenario in detail

vasicvuk commented:
Sure,

We have multiple Cloud native apps built with 30-50 microservices, each application uses some of the common services we call platform and some services specific to their product.

All of the microservices have their own charts and since we standardize the helm chart with internal generator tool we have a file named global which values from are applied for every microservice.

Other then that we have a folder for each team (owner of multiple microservice), and also an override which we use to optimize the helmfile for certain client were we implement product to. 

```
.
├── ...
├── base # Default configuration for the product
│   ├── global #  Global configuration
│   │   ├── values.yaml # Global values
│   │   └── 
│   ├── platform # Platform microservices
│   │   ├── service1.yaml # Platform microservice 1
│   │   ├── service2.yaml # Platform microservice 2
│   │   └── 
│   ├── app  # App microservices
│   │   ├── app1.yaml # App microservice 1
│   │   ├── app2.yaml # App microservice 2
│   │   └── 
│   └── 
├── override # Override configuration for the specific client 
│   ├── global #  Global override configuration
│   │   ├── values.yaml # Global values
│   │   └── 
│   ├── platform # Platform override microservices
│   │   ├── service1.yaml # Platform microservice 1
│   │   └── 
│   ├── app  # App override microservices
│   │   ├── app1.yaml # App microservice 1
│   │   └── 
│   └── 
└── ...
```

In override, the user can override only certain services, global values, or everything (it's up to the implementation team). Other than that we allow the option for an override file to begin with `_` and then the microservice will not be deployed at all even if it is defined in **base**.

We implemented everything from the description above, but currently, we need to have a list of folders in YAML so we know what folders are there (app, platform, third-party) and we want to automate also this by having an option to list folders from code too.

yxxhero commented:
@vasicvuk ok. Thanks very much. Perfect this PR according to your idea. I feel it is really useful.

vasicvuk commented:
@yxxhero, Thanks, I have changed the function to listDir and added unit tests for both `readDir` and `listDir`.

Since I am not really a GO developer, I need a little more time to understand how e2e part works, I am not sure if someone can jump in and help with e2e part? 

Thanks

vasicvuk commented:
@yxxhero  I managed to add e2e tests also :)

yxxhero commented on pkg/tmpl/context_funcs.go:
```diff
@@ -78,6 +87,14 @@ func (c *Context) createFuncMap() template.FuncMap {
+		}
+		funcMap["listDir"] = func(string) ([]string, error) {
+			var paths []string
+			return paths, DisableInsecureFeaturesErr
```
return nil, DisableInsecureFeaturesErr

vasicvuk replied:
Tnx, done

yxxhero commented on pkg/tmpl/context_funcs.go:
```diff
@@ -217,27 +232,35 @@ func (c *Context) ReadFile(filename string) (string, error) {
 	}
 
-	entries, err := os.ReadDir(contextPath)
+	entries, err := c.readDir(contextPath)
```
why change to c.readDir?

vasicvuk replied:
Because then i can mock it just like it was mocked in ReadFile function:

https://github.com/helmfile/helmfile/blob/047de0aa14bb7f02e790e9280c0ad31ce2dcb57b/pkg/tmpl/context_funcs.go#L227

vasicvuk replied:
For E2E tests it is using real os.ReadDir function and for tests it is using the mocked one then.

yxxhero replied:
ok

yxxhero commented on pkg/tmpl/context_funcs.go:
```diff
@@ -69,6 +70,14 @@ func (c *Context) createFuncMap() template.FuncMap {
 		}
+		funcMap["readDir"] = func(string) ([]string, error) {
+			var paths []string
+			return paths, nil
```
return nil, nil

vasicvuk replied:
I thought of that but I see that readFile for returns empty string. I can change it but I am then I guess this should be changed for readFile or I am wrong?

Thanks

yxxhero commented on pkg/tmpl/context_funcs.go:
```diff
@@ -69,6 +70,14 @@ func (c *Context) createFuncMap() template.FuncMap {
+			return paths, nil
+		}
+		funcMap["listDir"] = func(string) ([]string, error) {
+			var paths []string
```
return nil,nil

vasicvuk replied:
@yxxhero Are you sure, readFile function here returns empty string not nil?

yxxhero replied:
empty slice can be nil, but empty string can't

vasicvuk replied:
so maybe we should go with paths []string{}  instead of paths []string?

yxxhero replied:
return []string{}, nil?

vasicvuk replied:
@yxxhero sure, then it will be empty array as I understand

yxxhero replied:
@vasicvuk It's ok too.

vasicvuk replied:
Thanks :), I will modify it a little later when I am near the PC

yxxhero commented:
@mumoshu ping

mumoshu commented on docs/index.md:
```diff
@@ -679,6 +681,8 @@ You can use go's text/template expressions in `helmfile.yaml` and `values.yaml.g
 In addition to built-in ones, the following custom template functions are available:
 
 - `readFile` reads the specified local file and generate a golang string
+- `readDir` reads the files within provided directory path. (folders are excluded)
```
I prefer having only one function that reads a directory, named `readDir`, and let the caller decide what to do on each directory entry(=file info) by using e.g. `{{ if $fileEntry.IsDir }}`, so that we won't end up with two similar functions that a bit hard to remember 😄 
https://pkg.go.dev/io/ioutil#ReadDir

yxxhero replied:
good idea.

strainovic replied:
It sounds fine to me. The only unknown is if anyone already uses this undocumented function as this will be breaking change, right? Also, `isDir` has to be implemented, but it is easy to do.

vasicvuk replied:
Well if it returns an FileEntry object that I guess it isDir can be used from it, but still it will be a breaking change.

Maybe to refactor listDir to work that way and to mark existing readDir function as depracated @mumoshu ?

mumoshu replied:
Ahh, good catch! Indeed, we shouldn't introduce a breaking change.

How about adding `readDirEntries` backed by https://pkg.go.dev/os#ReadDir then? It sounds slightly more correct than `listDir`(it doesn't actually "list directories") and `listDirEntries` to me.

vasicvuk replied:
I agree with your proposal, I will try to refactor the code.

vasicvuk replied:
@mumoshu  I refactored the code as suggested

yxxhero commented on pkg/tmpl/context_funcs_test.go:
```diff
@@ -82,6 +84,94 @@ func TestReadFile(t *testing.T) {
+
+	expectedDirname := "sampleDirectory"
+	ctx := &Context{basePath: ".", readDir: func(dirname string) ([]fs.DirEntry, error) {
+		if dirname != expectedDirname {
```
@vasicvuk Please try to use `requires` library. Thanks very much

vasicvuk replied:
@yxxhero  Sorry, i don't understand. I see that there are require calls for test cooperation but this is a mock of a function, not a test case?

yxxhero commented:
https://github.com/helmfile/helmfile/blob/4e9b99d10e07c072287f108569c5d27f018b28e7/pkg/argparser/args_test.go#L8

vasicvuk commented:
@yxxhero  Should I change whole file then, as everything is using deepEqual currently?

yxxhero commented:
@vasicvuk whole file.

yxxhero commented on pkg/tmpl/context_funcs_test.go:
```diff
@@ -77,9 +78,91 @@ func TestReadFile(t *testing.T) {
+	}}
+
+	actual, err := ctx.ReadDir(expectedDirname)
+	if err != nil {
```
require.NoError?

vasicvuk commented:
@yxxhero  Done

yxxhero reviewed (approved):
LGTM

yxxhero commented:
@vasicvuk please fix go test error

vasicvuk commented:
@yxxhero  Can you help with identifying what the problem is, since i didn't change snapshot_test.go and it is showing this test as failure:

```
snapshot_test.go:126: helm push /home/runner/work/helmfile/helmfile/test/e2e/template/helmfile/raw-0.1.0.tgz oci://localhost:5000/myrepo: Error: failed to do request: Head "http://localhost:5000/v2/myrepo/raw/blobs/sha256:82d11697f6b3c7c5bcf619fb237648b2a8daee142dd0972f7f6244e72ea921e7": EOF
    snapshot_test.go:126: Unable to run helm push /home/runner/work/helmfile/helmfile/test/e2e/template/helmfile/raw-0.1.0.tgz oci://localhost:5000/myrepo: exit status 1
--- FAIL: TestHelmfileTemplateWithBuildCommand (10.77s)
    --- PASS: TestHelmfileTemplateWithBuildCommand/chart_need (6.33s)
    --- PASS: TestHelmfileTemplateWithBuildCommand/issue_2098_release_template_needs (0.07s)
    --- FAIL: TestHelmfileTemplateWithBuildCommand/oci_need (4.37s)
```

mumoshu commented:
Thanks for your efforts!
Probably that's a flaky test added via #239. Let me rerun it and see if the error disappears.

vasicvuk commented:
@mumoshu Yes, now they are passing :)
